AEP/aep-00004.md

```diff
@@ -0,0 +1,187 @@
+The recommended way would be to enforce listing tasks and results on sessions and adapt the GUI to match that requirement.
+To have a global overview, it would be necessary to add some session-wide statistics like the number of tasks completed and in error to enable the view at the session level, without the need to query all the clusters.
+The aggregation could be performed at regular interval on all opened sessions to reduce the number of sessions to query, as a closed session cannot have more tasks added.
+
+### Dynamic clusters
+
+The design impose no constraint on the addition of new clusters during operation.
+It would be a simple matter of allowing a new cluster to be chosen at session creation.
+
+The removal of a cluster is trickier.
+This would be possible only if all the sessions of the cluster are closed to avoid computation loss.
+
+If the monitoring of removed clusters is required, a dedicated database would be needed to store all the metadata of the removed clusters.
+When a RPC targets a removed cluster, this backup database could be requested instead.
+
+## Further considerations
+
+### Global monitoring
+
+Imposing to have a monitoring by session is simple enough, but is not really useful.
+A better approach would be to enable the monitoring spanning all clusters, and allow the listing of the tasks from all clusters.
+
+The main issue for that is the paginated listing of tasks or results.
+Indeed, it is not possible to know what pages of which cluster is required to get the n-th page of the global view.
+
+A way would be to introduce gRPC streaming for listing.
+With that, it would be possible for the Load Balancer to request all the clusters and merge their stream using the specified sort.
+
+It is not clear how this could be integrated into the Web GUI where the pages are mandatory for UX purposes.
+Either the GUI could materialize the whole stream when getting the pages, or a dedicated service could perform this materialization task.
+The latter approach would be more complex especially to avoid system overloads in the presence of many client, but we would entirely compatible with the current API.
+
+### On-demand clusters
+
+An extension of the Load Balancer could be to provision clusters on the fly.
+When a session is created, the Load Balancer could deploy a fresh cluster that will be destroyed when the session is deleted.
+
+The design of such a solution is more complex and could be considered later on.
+
+### Cluster field
+
+An alternative to the session ID field being used to load balance the request would be to add a cluster field to all the rpc which can be used to load balance the requests.
+It could also be used for monitoring instead of imposing a filter on the session.
+
+This change would be more intrusive API-wise.
+
+# Rejected idea
+
+## Deterministic binding
+
+The first idea we had for a stateless Load Balancer was to avoid discovery altogether and have a deterministic binding between session and cluster.
```

Comment:
```suggestion
The first idea for a stateless Load Balancer was to avoid discovery altogether and have a deterministic binding between session and cluster.
```